src/KafkaW/MetaDataQueryImpl.h

```diff
@@ -0,0 +1,131 @@
+      KafkaHandle::create(Conf.get(), ErrorStr));
+  if (KafkaConsumer == nullptr) {
+    throw MetadataException("Unable to create kafka handle.");
+  }
+  return KafkaConsumer;
+}
+
+template <class KafkaHandle>
+std::vector<std::pair<int, int64_t>>
+getOffsetForTimeImpl(std::string const &Broker, std::string const &Topic,
+                     std::vector<int> const &Partitions, time_point Time,
+                     duration TimeOut) {
+  auto Handle = getKafkaHandle<KafkaHandle, RdKafka::Conf>(Broker);
+  auto UsedTime = std::chrono::duration_cast<std::chrono::milliseconds>(
+                      Time.time_since_epoch())
+                      .count();
+
+  std::vector<std::unique_ptr<RdKafka::TopicPartition>> TopicPartitions;
+  std::vector<RdKafka::TopicPartition *> TopicPartitionsRaw;
+  for (const auto &PartitionId : Partitions) {
+    auto CTopicPartition = std::unique_ptr<RdKafka::TopicPartition>(
+        RdKafka::TopicPartition::create(Topic, PartitionId, UsedTime));
+    TopicPartitionsRaw.emplace_back(CTopicPartition.get());
+    TopicPartitions.push_back(std::move(CTopicPartition));
+  }
+
+  auto TimeOutInMs =
+      std::chrono::duration_cast<std::chrono::milliseconds>(TimeOut).count();
+
+  auto ReturnCode = Handle->offsetsForTimes(TopicPartitionsRaw, TimeOutInMs);
+  if (ReturnCode != RdKafka::ERR_NO_ERROR) {
+    throw MetadataException("Failed to query broker for offset corresponding "
+                            "to timestamp. Error code was: " +
+                            std::to_string(ReturnCode));
+  }
+  std::vector<std::pair<int, int64_t>> ReturnSet;
+  for (const auto &CTopicPartition : TopicPartitions) {
+    ReturnSet.emplace_back(std::make_pair(CTopicPartition->partition(),
+                                          CTopicPartition->offset()));
+  }
+  return ReturnSet;
+}
+
+template <class KafkaHandle, class KafkaTopic>
+std::vector<int> getPartitionsForTopicImpl(std::string const &Broker,
+                                           std::string const &Topic,
+                                           duration TimeOut) {
+  auto Handle = getKafkaHandle<KafkaHandle, RdKafka::Conf>(Broker);
+  std::string ErrorStr;
+  auto TopicObj = std::unique_ptr<RdKafka::Topic>(
+      KafkaTopic::create(Handle.get(), Topic, nullptr, ErrorStr));
+  auto TimeOutInMs =
+      std::chrono::duration_cast<std::chrono::milliseconds>(TimeOut).count();
+  RdKafka::Metadata *MetadataPtr{nullptr};
+  auto ReturnCode =
+      Handle->metadata(true, TopicObj.get(), &MetadataPtr, TimeOutInMs);
+  if (ReturnCode != RdKafka::ERR_NO_ERROR) {
+    throw MetadataException(
+        "Failed to query broker for available partitions. Error code was: " +
+        std::to_string(ReturnCode));
+  }
+  auto TopicMetaData = findKafkaTopic(Topic, MetadataPtr);
+  std::vector<int> ReturnVector;
+  for (auto const &Partition : *TopicMetaData->partitions()) {
+    ReturnVector.push_back(Partition->id());
+  }
+  delete MetadataPtr;
```

Comment:
The six lines of code above have complicated lifetimes, I would suggest trying to move 100 to 103 into `findKafkaTopic` and just return the partition vector from there.
Then rename `findKafkaTopic` to reflect this.

MetadataPtr can be a unique_ptr?

Reply:
About `unique_ptr`: No it can not. `std::unique_ptr::get()` is `const`.